### PR TITLE
Added daemon related functions in heal ops

### PR DESCRIPTION
**Pull request by aujjwal-redhat**
Added:
1. is_shd_daemon_running
2. is_shd_daemonized
3. get_heal_info_summary
4. get_self_heal_daemon_pid

Fixes: #700

Signed-off-by: Ayush Ujjwal <aujjwal@redhat.com>

### CODE REVIEW

**Comment by aujjwal-redhat** [comment]
```js

 test_heal_ops
 
+-------------+-------------+--------------------+
| Volume Type | Test Result |  Time taken (sec)  |
+-------------+-------------+--------------------+
|     rep     |     PASS    | 145.49896430969238 |
+-------------+-------------+--------------------+

 vol_destroy_test
 
+-------------+-------------+-------------------+
| Volume Type | Test Result |  Time taken (sec) |
+-------------+-------------+-------------------+
|     rep     |     PASS    | 9.297699213027954 |
+-------------+-------------+-------------------+
```

File: common/ops/gluster_ops/heal_ops.py [hunk 1]
```diff
@@ -129,6 +129,46 @@ def get_heal_info(self, node: str, volname: str) -> list:
 
         return ret['msg']['healInfo']['bricks']['brick']
 
+    def get_heal_info_summary(self, node: str, volname: str) -> dict:
+        """
+        From the xml output of heal info command  get heal info summary
+        i.e Bricks and it's corresponding number of entries, status.
+
+        Args:
+            mnode : Node on which commands are executed
```
Comment (nik-redhat) [inline common/ops/gluster_ops/heal_ops.py:138]: nit: node

File: common/ops/gluster_ops/heal_ops.py [hunk 2]
```diff
@@ -291,3 +331,129 @@ def is_volume_in_split_brain(self, node: str, volname: str) -> bool:
...
+        _rc = True
+
+        if not isinstance(nodes, list):
+            nodes = [nodes]
+        cmd = r"pgrep -f glustershd | grep -v ^$$\$"
+        self.logger.info(f"Running {cmd} on nodes {nodes}")
+        results = self.execute_abstract_op_multinode(cmd, nodes)
+
+        for result in results:
+            if result['error_code'] == 0:
```
Comment (nik-redhat) [inline common/ops/gluster_ops/heal_ops.py:363]: If you are checking it here, pass `exep=False` above to abstract ops.

File: common/ops/gluster_ops/heal_ops.py [hunk 3]
```diff
@@ -291,3 +331,129 @@ def is_volume_in_split_brain(self, node: str, volname: str) -> bool:
...
+                   False Otherwise
+        """
+        counter = 0
+        flag = 0
+
+        if not isinstance(nodes, list):
+            nodes = [nodes]
+
+        while counter < timeout:
+            ret = self.get_self_heal_daemon_pid(nodes)
```
Comment (nik-redhat) [inline common/ops/gluster_ops/heal_ops.py:409]: Shouldn't the return value be received like `ret, _ = self.get....`?

File: common/ops/gluster_ops/heal_ops.py [hunk 4]
```diff
@@ -291,3 +331,129 @@ def is_volume_in_split_brain(self, node: str, volname: str) -> bool:
...
+
+            if not ret:
+                self.logger.info("Retry after 3 sec to get"
+                                 " self-heal daemon process.....")
+                sleep(3)
+                counter = counter + 3
+            else:
+                flag = 1
```
Comment (nik-redhat) [inline common/ops/gluster_ops/heal_ops.py:417]: Is another `flag` variable needed? We can use `ret` later on as well, right?

File: common/ops/gluster_ops/heal_ops.py [hunk 5]
```diff
@@ -291,3 +331,129 @@ def is_volume_in_split_brain(self, node: str, volname: str) -> bool:
...
+                break
+
+        if not flag:
+            self.logger.error("Either No self heal daemon process found "
+                              "or more than one self heal daemon process"
+                              f" found even after {timeout/60.0} minutes")
+            return False
```
Comment (nik-redhat) [inline common/ops/gluster_ops/heal_ops.py:424]: Let's change the last return to `return ret` after following the comments above. That should do as well.

File: common/ops/gluster_ops/heal_ops.py [hunk 6]
```diff
@@ -291,3 +331,129 @@ def is_volume_in_split_brain(self, node: str, volname: str) -> bool:
...
+        else:
+            self.logger.info("Single self heal daemon process on"
+                             f" all nodes {nodes}")
+        return True
+
+    def is_shd_daemon_running(self, mnode: str,
```
Comment (nik-redhat) [inline common/ops/gluster_ops/heal_ops.py:430]: For consistency, keep it `node`.

File: common/ops/gluster_ops/heal_ops.py [hunk 7]
```diff
@@ -291,3 +331,129 @@ def is_volume_in_split_brain(self, node: str, volname: str) -> bool:
...
+                              node: str, volname: str) -> bool:
```
Comment (nik-redhat) [inline common/ops/gluster_ops/heal_ops.py:431]: And make this `node` -> `server`

File: common/ops/gluster_ops/heal_ops.py [hunk 8]
```diff
@@ -291,3 +331,129 @@ def is_volume_in_split_brain(self, node: str, volname: str) -> bool:
...
+        """
+        Verifies whether the shd daemon is up and running on a
+        particular node by checking the existence of shd pid
+        and parsing the get volume status output.
+
+        Args:
+            mnode (str): The first node in servers list
+            node (str): The node to be checked for whether
```
Comment (nik-redhat) [inline common/ops/gluster_ops/heal_ops.py:439]: As per the above change.

**Review by nik-redhat — CHANGES_REQUESTED** [review]

File: common/ops/gluster_ops/heal_ops.py (hunk 1 above)
Comment (aujjwal-redhat) [reply common/ops/gluster_ops/heal_ops.py:138]: Ack

File: common/ops/gluster_ops/heal_ops.py (hunk 2 above)
Comment (aujjwal-redhat) [reply common/ops/gluster_ops/heal_ops.py:363]: Ack

File: common/ops/gluster_ops/heal_ops.py (hunk 3 above)
Comment (aujjwal-redhat) [reply common/ops/gluster_ops/heal_ops.py:409]: Yeah, true. Actually,

File: common/ops/gluster_ops/heal_ops.py (hunk 4 above)
Comment (aujjwal-redhat) [reply common/ops/gluster_ops/heal_ops.py:417]: Yeah

File: common/ops/gluster_ops/heal_ops.py (hunk 7 above)
Comment (aujjwal-redhat) [reply common/ops/gluster_ops/heal_ops.py:431]: ok

File: common/ops/gluster_ops/heal_ops.py (hunk 5 above)
Comment (aujjwal-redhat) [reply common/ops/gluster_ops/heal_ops.py:424]: yeah that's better

File: common/ops/gluster_ops/heal_ops.py (hunk 6 above)
Comment (aujjwal-redhat) [reply common/ops/gluster_ops/heal_ops.py]: Ack

File: common/ops/gluster_ops/heal_ops.py [hunk 9]
```diff
@@ -291,3 +331,127 @@ def is_volume_in_split_brain(self, node: str, volname: str) -> bool:
...
+        if not isinstance(nodes, list):
+            nodes = [nodes]
+        cmd = r"pgrep -f glustershd | grep -v ^$$\$"
+        self.logger.info(f"Running {cmd} on nodes {nodes}")
+        results = self.execute_abstract_op_multinode(cmd, nodes, False)
+
+        for result in results:
+            if result['error_code'] == 0:
+                if len(result['msg']) == 1:
+                    if not (result['msg'][0]).strip().rstrip("\n"):
```
Comment (nik-redhat) [inline common/ops/gluster_ops/heal_ops.py:365]: Are both of the `strip` functions needed?

Comment (aujjwal-redhat) [reply common/ops/gluster_ops/heal_ops.py:365]: Just considering the worst case like "             xyz\n          " XD

Comment (nik-redhat) [reply common/ops/gluster_ops/heal_ops.py:365]: Ack!

**Review by nik-redhat — APPROVED** [review]

**Pull request merged**